### PR TITLE
Add more `idaklu` solver options

### PR DESCRIPTION
**Pull request by MarcBerliner**
# Description

Adds more user-configurable options to the IDA solver following the list in the IDA guide.

Some code cleanup to enforce early returns.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by valentinsulzer — COMMENTED** [review]
Looks good to me, though I'm not familiar enough with C++ to properly review those files

**Comment by kratman** [comment]
@valentinsulzer I will review the C++ once the tests are passing

**Comment by kratman** [comment]
@MarcBerliner If you want the codecov failure to go away, rebasing should make the PR start using the correct commit. Since the review has not started yet, it is still fine to force push

**Comment by MarcBerliner** [comment]
@kratman It looks like the final real failure is with the static code analysis we discussed with the `Options` struct. Do you have a recommendation on how to proceed?

**Comment by kratman** [comment]
I will look at that with the code review

**Review by kratman — CHANGES_REQUESTED** [review]
Partial review

**Comment by kratman** [comment]
@MarcBerliner The static analysis is correct. You are not using those variables. It is related to the scope issue that I mentioned in our call earlier:
```c++
CasadiSolverOpenMP::CasadiSolverOpenMP(
  np_array atol_np,
  double rel_tol,
  np_array rhs_alg_id,
  int number_of_parameters,
  int number_of_events,
  int jac_times_cjmass_nnz,
  int jac_bandwidth_lower,
  int jac_bandwidth_upper,
  std::unique_ptr<CasadiFunctions> functions_arg,
  const Options &options
) :
  atol_np(atol_np),
  rhs_alg_id(rhs_alg_id),
  number_of_states(atol_np.request().size),
  number_of_parameters(number_of_parameters),
  number_of_events(number_of_events),
  jac_times_cjmass_nnz(jac_times_cjmass_nnz),
  jac_bandwidth_lower(jac_bandwidth_lower),
  jac_bandwidth_upper(jac_bandwidth_upper),
  functions(std::move(functions_arg)),
  options(options)
```
At the bottom the member variable `options` has the same name as the input reference `const Options &options`

Then in your code you are using the following:
```c++
IDASetMaxOrd(ida_mem, options.max_order_bdf);
```
This is ambiguous due to the name options being in multiple scopes. This would be corrected by doing:
```c++
IDASetMaxOrd(ida_mem, this->options.max_order_bdf);
```
In C++ the `this->` is effectively `self.` in python. When there is not scope ambiguity, then the `this->` can be omitted. So what you wrote is basically the following:
```python
self.options = options
IDASetMaxOrd(ida_mem, options.max_order_bdf)
```
instead of 
```python
self.options = options
IDASetMaxOrd(ida_mem, self.options.max_order_bdf)
```
So you never use the variables you added to the struct, you simply take those settings from the input variable.

If those variables are not needed anywhere else then you can look into inheriting from Options to make a new IKAOptions or something along those lines.

**Comment by kratman** [comment]
As a followup for making this easier, this problem would also be avoided by putting all the settings code into a method:
```c++
void CasadiSolverOpenMP::SetOptionsForIDA()
{
  // Maximum order of the linear multistep method
  IDASetMaxOrd(ida_mem, options.max_order_bdf);

  // Maximum number of steps to be taken by the solver in its attempt to reach
  // the next output time
  IDASetMaxNumSteps(ida_mem, options.max_num_steps);

  // Initial step size
  IDASetInitStep(ida_mem, RCONST(options.dt_init));

  // Maximum absolute step size
  IDASetMaxStep(ida_mem, RCONST(options.dt_max));

  // Maximum number of error test failures in attempting one step
  IDASetMaxErrTestFails(ida_mem, options.max_error_test_failures);

  // Maximum number of nonlinear solver iterations at one step
  IDASetMaxNonlinIters(ida_mem, options.max_nonlinear_iterations);

  // Maximum number of nonlinear solver convergence failures at one step
  IDASetMaxConvFails(ida_mem, options.max_convergence_failures);

  // Safety factor in the nonlinear convergence test
  IDASetNonlinConvCoef(ida_mem, RCONST(options.nonlinear_convergence_coefficient));

  // Suppress algebraic variables from error test
  IDASetSuppressAlg(ida_mem, options.suppress_algebraic_error);

  // Positive constant in the Newton iteration convergence test within the
  // initial condition calculation
  IDASetNonlinConvCoefIC(ida_mem, RCONST(options.nonlinear_convergence_coefficient_ic));

  // Maximum number of steps allowed when icopt=IDA_YA_YDP_INIT in IDACalcIC
  IDASetMaxNumStepsIC(ida_mem, options.max_num_steps_ic);

  // Maximum number of the approximate Jacobian or preconditioner evaluations
  // allowed when the Newton iteration appears to be slowly converging
  IDASetMaxNumJacsIC(ida_mem, options.max_number_jacobians_ic);

  // Maximum number of Newton iterations allowed in any one attempt to solve
  // the initial conditions calculation problem
  IDASetMaxNumItersIC(ida_mem, options.max_number_iterations_ic);

  // Maximum number of linesearch backtracks allowed in any Newton iteration
  //, when solving the initial conditions calculation problem
  IDASetMaxBacksIC(ida_mem, options.max_linesearch_backtracks_ic);

  // Turn off linesearch
  IDASetLineSearchOffIC(ida_mem, options.linesearch_off_ic);
}
```
This method takes no arguments, so all the data must be taken from the options. However, if those options are never used again, then maybe they do not need to be stored.

It is also not just your code that has this issue. It looks like right below your code the same mistake exists:
```c++
  // set events
  IDARootInit(ida_mem, number_of_events, events_casadi);
  void *user_data = functions.get();
  IDASetUserData(ida_mem, user_data);

  // specify preconditioner type
  precon_type = SUN_PREC_NONE;
  if (options.preconditioner != "none") {
    precon_type = SUN_PREC_LEFT;
  }
```
Here options and number_of_events are also taken from the arguments, not the member data of the class.

File: pybamm/solvers/idaklu_solver.py
Comment (MarcBerliner) [reply pybamm/solvers/idaklu_solver.py:675]: I didn't write this block, I just moved up the `pybamm.SolverError("idaklu solver failed")` to get rid of the if/else block. But yeah this is unnecessary.

File: pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp
Comment (kratman) [inline pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp:242]:
```suggestion
  if (number_of_parameters > 0) {
      IDASensFree(ida_mem);
  }
```

File: pybamm/solvers/c_solvers/idaklu/options.cpp
Comment (kratman) [inline pybamm/solvers/c_solvers/idaklu/options.cpp:149]:
```suggestion
{}

```

**Review by kratman — CHANGES_REQUESTED** [review]
Mostly minor stuff. I will double check the Codacy failures tomorrow.

File: pybamm/solvers/idaklu_solver.py
Comment (MarcBerliner) [reply pybamm/solvers/idaklu_solver.py:108]: With `dt_init` and `dt_max`, a value of `0.0` uses the default value. I'll add a note to explain this.

File: pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp
Comment (kratman) [inline pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp:223]: Not completely necessary since it is most likely optimized out, but you can simplify this since the flag is not needed later
```suggestion
  CheckErrors(IDASetLinearSolver(ida_mem, LS, J));
```

File: pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp
Comment (kratman) [inline pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp:651]: This can be made const since it never changes
```suggestion
void CasadiSolverOpenMP::CheckErrors(int const & flag) {
```

File: pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.hpp
Comment (kratman) [inline pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.hpp:157]: pairs with the previous change
```suggestion
  void CheckErrors(int const & flag);
```

File: pybamm/solvers/idaklu_solver.py
Comment (kratman) [inline pybamm/solvers/idaklu_solver.py:110]: We should be able to make these comments on the same line to keep it from looking like commented out code
```suggestion
            "dt_init": 0.0,  # The solver default is used if this is left at zero
            "dt_max": 0.0,  # The solver default is used if this is left at zero
```

File: pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp
Comment (kratman) [reply pybamm/solvers/c_solvers/idaklu/CasadiSolverOpenMP.cpp:223]: Feel free to just make a cleanup ticket so this PR can advance. This is not necessary for this PR to be approved

**Comment by kratman** [comment]
I am finishing up double checking codacy, so this should be ready soon

**Comment by kratman** [comment]
As far as I can tell the codacy is a false alarm at this point. It even mentions the old struct names which do not exist anymore. I manually checked what was there, and the values appear to be used. Considering that there are some scope issues in the C++ code and other PRs are heavily changing the code, it is probably not worth suppressing the warnings

**Review by kratman — APPROVED** [review]

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Hi All, sorry to be late to this party, I should have caught this one earlier. @jsbrittain has had a PR up for a while now (which is almost ready to be merged) https://github.com/pybamm-team/PyBaMM/pull/4199, which substantially refactors the C++ solver code. Can we make sure that any changes to the C++ solver are rebased on top of this, so they can be merged after we merge in #4199?

**Comment by martinjrobins** [comment]
Update: I've had a look at #4199 now and this is good to be merged in. @MarcBerliner @kratman: are you happy if we merge #4199 in and then you can rebase/merge ontop of it? #4199 doesn't touch the options class (other than move it to a new folder) so cross fingers this should be straightforward, but I'm happy to help out with this on monday if needed.

**Comment by MarcBerliner** [comment]
Let's merge in #4199 first. I will be away this week, but afterwards I can update this PR

**Comment by MarcBerliner** [comment]
Closing due to merging https://github.com/pybamm-team/PyBaMM/pull/4282

**Pull request closed without merging**
